ui/src/router.tsx

```diff
@@ -0,0 +1,94 @@
+import Button from "@mui/material/Button";
+import Edit from "edit";
+import { PathError } from "hooks";
+import Overview from "overview";
+import { ErrorBoundary } from "react-error-boundary";
+import {
+  Route,
+  Switch,
+  useHistory,
+  useParams,
+  useRouteMatch,
+} from "react-router-dom";
+import { UnderlaySelect } from "underlaySelect";
+import { Datasets } from "./datasets";
+
+const pages = {
+  "/": UnderlaySelect,
+  "/:underlayName": Datasets,
+  "/:underlayName/cohorts/:cohortId": Overview,
+  "/:underlayName/cohorts/:cohortId/edit/:groupId/:criteriaId": Edit,
+};
+
+export function AppRouter() {
+  return (
+    <ErrorBoundary FallbackComponent={NotFound}>
+      <Switch>
+        {Object.entries(pages).map(([page, component]) => (
+          <Route exact key={page} path={page} component={component} />
+        ))}
+        <Route path="/">
+          <NotFound error={new PathError("Invalid URL.")} />
+        </Route>
+      </Switch>
+    </ErrorBoundary>
+  );
+}
+
+// useBackUrl returns the second longest matching page, assuming that the
```

Comment:
I can imagine cases where this logic won't work as expected, for example if you're in a flow between siblings. Are there issues using the history more explicitly for the current usages?

Alternatively, name this something like `useParentUrl()`

Reply:
So I had originally built a more complicated version that explicitly specified the back URLs for each page but it was more complicated for no benefit. I'm not planning any behaviour that wouldn't fit this model but this can be revisited if something does come up.

Using history explicitly is more of a UX decision. In my opinion, the back button on the page is more of an "up" button allowing the user to get to the parent page. There's no reason to duplicate the browser's back button and having it work this way is actually helpful in situations where there is more complicated navigation since it guarantees a way to get to the parent even if that's not where the user came from.

Reply:
How about something like `useParentUrl()`, then? `useBackUrl()` seems overly prescriptive on where you're expecting this to be used the app, and inaccurate based on what it's doing.

Reply:
Done.